### PR TITLE
Adding auto test for ANM

### PR DESCRIPTION
**Pull request by ErdunGAO**
**Updates**
- Remove `print` commands and add assertions.
- Add more simulated functions and save the data.
- Fix some typos.

**Test plan**
`python -m unittest TestANM.py      # should pass`

image

### CODE REVIEW

**Review by tofuwen — COMMENTED** [review]
Thanks for the great work!

Some general comments:
1. please refer to https://github.com/cmu-phil/causal-learn/pull/51 to check the discussion on how to make sure the code logic is unchanged after each updates. 

I don't know whether the method there will apply to ANM or not. But asserting p_value > 0.05 seems too weak.

2. please add a test plan section in your PR. You can refer to https://github.com/cmu-phil/causal-learn/pull/51. Note please add a screenshot of your test pass message in the terminal in the test plan as well.

File: tests/TestANM.py [hunk 1]
```diff
@@ -8,44 +8,49 @@
...
+        p_value_1, _ = anm.cause_or_effect(x, y_1)
+        self.assertTrue(p_value_1 > 0.05)
+
+        # simulated data y = 3*exp(x) + e
+        y_2 = 3 * np.exp(x) + e
+        p_value_2, _ = anm.cause_or_effect(x, y_2)
+        self.assertTrue(p_value_2 > 0.05)
+
+        # simulated data y = 2 * Sigmoid(x) + e
+        y_3= 2/(1+np.exp(-x)) + e
```
[the remark this answers is not in the document]
Comment (ErdunGAO) [reply tests/TestANM.py:40]: Thanks, the space issue would be fixed next time.

File: tests/TestANM.py [hunk 2]
```diff
@@ -8,44 +8,49 @@
...
-        df = pd.read_csv('TestData/pair0001.txt', sep=' ', header=None)
-        dataset = df.to_numpy()
+
+        dataset = np.loadtxt('TestData/pair0001.txt')
         anm = ANM()
-        n = dataset.shape[0]
-        p_value_foward, p_value_backward = anm.cause_or_effect(dataset[:, 0].reshape(n, 1), dataset[:, 1].reshape(n, 1))
-        print('pvalue for x->y is {:.4f}'.format(p_value_foward))
-        print('pvalue for y->x is {:.4f}'.format(p_value_backward))
+        p_value_forward, _ = anm.cause_or_effect(dataset[:, 0].reshape(-1, 1), dataset[:, 1].reshape(-1, 1))
```
[the remark this answers is not in the document]
Comment (ErdunGAO) [reply tests/TestANM.py:49]: Thanks, the p_value_backward will be checked in the next version.

File: tests/TestANM.py [hunk 3]
```diff
@@ -8,44 +8,49 @@
...
+        self.assertTrue(p_value_forward > 0.05)
 
 
 if __name__ == '__main__':
```
[the remark this answers is not in the document]
Comment (ErdunGAO) [reply tests/TestANM.py:53]: Thanks, this will be well changed.

**Comment by tofuwen** [comment]
@ErdunGAO You should fix the issues in this PR, and update your current PR (instead of doing it in next PR). This is why we need code review --- to guarantee the code quality is good before pushing the code into main trunk.

If you are not sure how to do it, let me know. Actually this should be really easy --- just commit to your forked repo with your new changes, and then create Pull Request.

**Comment by ErdunGAO** [comment]
> 

1. In the new version, `p_value_th` is set to `0.1`. The specific `p_value` is related to the distribution of the simulated data.
2. The **test plan** section is added with the screenshot of the testing results.

Please let me know if there is any further concern. Thanks!

**Review by tofuwen — COMMENTED** [review]
Thanks for the great work! I have some nits that needs to be addressed.

BTW, I remembered that we discussed to use Haoyue's method to make sure our code logic doesn't change? 

See https://github.com/cmu-phil/causal-learn/pull/53 as another example.

In your specific case, you can do some easy things like: compute the current p_values using current code for 'TestData/pair0001.txt', and store it. And make sure the code can return the same result as your stored results. In this way, we can make sure when making future changes, if the code logic get changed, we can detect it.

File: tests/TestANM.py [hunk 4]
```diff
@@ -8,44 +8,49 @@
...
 from causallearn.search.FCMBased.ANM.ANM import ANM
 
 
 class TestANM(unittest.TestCase):
-    def syn_data(self, b, n):
-        # x = np.abs(np.random.randn(n, 1) * np.sign(np.random.randn(n, 1)))
-        x = np.random.randn(n, 1)
+
+    # Test ANM by some simulated data
+    p_value_th = 0.1
```
[the remark this answers is not in the document]
Comment (ErdunGAO) [reply tests/TestANM.py:18]: Thanks, this issue has been solved in the new version.

**Comment by ErdunGAO** [comment]
> Thanks for the great work! I have some nits that needs to be addressed.
> 
> BTW, I remembered that we discussed to use Haoyue's method to make sure our code logic doesn't change?
> 
> See #53 as another example.
> 
> In your specific case, you can do some easy things like: compute the current p_values using current code for 'TestData/pair0001.txt', and store it. And make sure the code can return the same result as your stored results. In this way, we can make sure when making future changes, if the code logic get changed, we can detect it.

Thanks for your kind suggestion! In the new version, simulated data have been well saved to ensure that consistent results can be obtained when meeting future changes.

**Comment by tofuwen** [comment]
Sorry for not making myself clear. 

Do you have some time today to meet to discuss?

**Comment by ErdunGAO** [comment]
>

**Comment by ErdunGAO** [comment]
> Sorry for not making myself clear.
> 
> Do you have some time today to meet to discuss?

Thanks a lot for your kind explanation! 
- The data simulation codes are well included in the new version. 
- We have saved the results for the fixed datasets, which can be recalled and compared to check future changes.

**Review by tofuwen — APPROVED** [review]
Thanks for the great work, this is awesome!!!

I think this PR is ready to be merged. cc @kunwuz

**Pull request merged**